src/adagrams.js

```diff
@@ -1,15 +1,132 @@
+
+  // creates letter bank with num of letters equal to letterPool values
+  for (let [key,value] of Object.entries(letterPool)) {
+    for (value in range(value)) {
+      letterBank.push(key);
+    }
+  }
+
+  // draws 10 random letters from letterBank
+  while (returnList.length < 10) {
+    let letter = letterBank[Math.floor(Math.random()* letterBank.length)];
+    returnList.push(letter);
+    // grabs index of letter from letterBank and removes that letter at the index
+    let index = letterBank.indexOf(letter);
+    letterBank.splice(index,1);
+  }
+
+  return returnList;
 };
 
-export const usesAvailableLetters = (input, lettersInHand) => {
-  // Implement this method for wave 2
+export const usesAvailableLetters = (word, letterBank) => {
+  let letters = letterBank.copyWithin();
+
+  const checkWord = []
+  for (let letter of word.toUpperCase()) {
+    // checks if letter in word is in letterBank
+    if (letters.includes(letter)) {
+      // adds letter to checkWord list
+      checkWord.push(letter);
+      // removes letter from letterBank so multiples can't be used
+      let index = letters.indexOf(letter);
+      letters.splice(index,1);
+    }
+  }
+
+  // if checkWord is the same length as input word, then the word uses available letters
+  if (checkWord.length == word.length) {
+    return true;
+  } else {
+    return false;
+  }
 };
 
 export const scoreWord = (word) => {
-  // Implement this method for wave 3
+  const scoreDict = {
+    'A': 1, 'B': 3, 'C': 3, 'D': 2, 'E': 1, 'F': 4, 
+    'G': 2, 'H': 4, 'I': 1, 'J': 8, 'K': 5, 'L': 1, 
+    'M': 3, 'N': 1, 'O': 1, 'P': 3, 'Q': 10, 'R': 1, 
+    'S': 1, 'T': 1, 'U': 1, 'V': 4, 'W': 4, 'X': 8, 
+    'Y': 4, 
+    'Z': 10
+  };
+
+  // calculates score by iterating through each letter in word and adding its value from scoreDict
+  let score = 0;
+  word.toUpperCase().split('').forEach((letter) => {
+    score += scoreDict[letter];
+  })
+
+  // adds 8 bonus points if word length is >= 7 and <= 10
+  if (word.length >= 7 && word.length <= 10) {
+    score += 8;
+  }
+
+  return score
 };
 
+
 export const highestScoreFrom = (words) => {
-  // Implement this method for wave 4
-};
+  // creates dictionary of scores for each word
+  let dictOfScores = {};
+  for (let word of words) {
+    let wordScore = scoreWord(word);
+    dictOfScores[word] = wordScore;
+  }
+
+  // creates list of words with highest score
+  let highestScore = Math.max.apply(null, Object.values(dictOfScores));
+  let maxScoreList = [];
+  for (let [key, value] of Object.entries(dictOfScores)) {
+    if (value === highestScore) {
+      maxScoreList.push(key);
+    }
+  }
+  
+  // returns highest score if length of list is only 1
+  if (maxScoreList.length === 1) {
+    return {word: maxScoreList[0], score: highestScore}
+  }
+
+  // gets max and min length word of highest scoring words
+  let maxLenWord = maxScoreList[0];
+  let minLenWord = maxLenWord;
+  for (let word of maxScoreList) {
```

Comment:
This works but you could simplify this by creating one variable to hold the high scoring data. Using two variables takes up more memory and most primitives (booleans, numbers, null, undefined, etc.) are constant (O(1)) space. This is because, no matter if true or false, or 1 or 1,000, the amount of space consumed stays constant. Reference types (arrays and objects) — Reference types require O(n) space as well. ’n’ being the length for an array, or in the case of an object, ’n’ being the amount of keys. Reference types are O(n) because the more elements in an array, or the more keys in an object, the more space they consume.

The `maxScoreList` array does have a space complexity of 0(n), so does the `dictOfScores` object. You are also using a `highestScore` variable which holds the result of calling `Math.max`, `apply()` and `Object.values()`. Your space complexity is also increased by the `maxLenWord` and `minLenWord` variables, consider checking if the current word is shorter than the current highest scoring word 👍🏾 

Consider also creating a data structure to hold both the winning word and score together and setting that value to falsey and then reassigning based on some condition. Sort of like the `get_highest_rated` PSE 😄